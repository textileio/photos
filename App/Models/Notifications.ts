import { pb } from '@textile/react-native-sdk'

interface BaseNotification {
  id: string
  date: Date
  actor: string
  block: string
  body: string
  read: boolean
  username: string
  avatar: string
}

export type InviteReceivedNotification = BaseNotification & {
  threadName: string
  type: pb.Notification.Type.INVITE_RECEIVED
}

export type AccountPeerJoinedNotification = BaseNotification & {
  type: pb.Notification.Type.ACCOUNT_PEER_JOINED
}

export type PeerJoinedNotification = BaseNotification & {
  threadId: string
  threadName: string
  type: pb.Notification.Type.PEER_JOINED
}

export type PeerLeftNotification = BaseNotification & {
  threadId: string
  threadName: string
  type: pb.Notification.Type.PEER_LEFT
}

export type MessageAddedNotification = BaseNotification & {
  threadId: string
  threadName: string
  target: string
  type: pb.Notification.Type.MESSAGE_ADDED
}

export type FilesAddedNotification = BaseNotification & {
  threadId: string
  threadName: string
  target: string
  type: pb.Notification.Type.FILES_ADDED
}

export type CommentAddedNotification = BaseNotification & {
  threadId: string
  threadName: string
  target: string
  type: pb.Notification.Type.COMMENT_ADDED
}

export type LikeAddedNotification = BaseNotification & {
  threadId: string
  threadName: string
  target: string
  type: pb.Notification.Type.LIKE_ADDED
}

export type Notification =
  InviteReceivedNotification |
  AccountPeerJoinedNotification |
  PeerJoinedNotification |
  PeerLeftNotification |
  MessageAddedNotification |
  FilesAddedNotification |
  CommentAddedNotification |
  LikeAddedNotification
