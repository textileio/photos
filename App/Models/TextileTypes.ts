// BEGIN API Types

// private enums to enforce strict id types
// read here: https://basarat.gitbooks.io/typescript/docs/tips/nominalTyping.html#using-enums

import {IProcessingImageProps} from '../Components/ProcessingImage'
import { ThreadFilesInfo } from '@textile/react-native-sdk'

export interface PhotoType {
  type: 'photo',
  photo: ThreadFilesInfo
  id: string
}

export interface ProcessingItemType {
  type: 'processingItem',
  photo: IProcessingImageProps
  id: string
}

export type IPhotoGridType = PhotoType | ProcessingItemType

export interface DeepLinkData {
  readonly href: string,
  readonly protocol: string,
  readonly host: string,
  readonly hostname: string,
  readonly port: string,
  readonly path: string,
  readonly search: string
  readonly hash: string
}

export interface ILocalPhotoResult {
  assetId: string,
  creationDate: string,
  modificationDate: string,
  orientation: number,
  path: string,
  uri: string,
  canDelete: boolean
}
