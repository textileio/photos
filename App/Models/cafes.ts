import Config from 'react-native-config'
import { Buffer } from 'buffer'

export interface Cafe {
  name: string
  peerId: string
  token: string
}

const cafesBase64 = Config.RN_TEXTILE_CAFES_JSON
const cafesString = new Buffer(cafesBase64, 'base64').toString()

export const cafes: Cafe[] = JSON.parse(cafesString)

export const cafesMap = cafes.reduce(
  (acc, current) => {
    return { ...acc, [current.peerId]: current }
  },
  {} as { [key: string]: Cafe | undefined }
)
