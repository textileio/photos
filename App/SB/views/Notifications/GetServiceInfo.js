export default function (service: string): {[index: string]: string} | undefined {
  switch (service) {
    case 'backgroundLocation':
      return {
        title: 'Wake with location',
        subtitle: 'More timely notifications',
        details: 'Background location allows Textile to wake up periodically to check for updates on your peer-to-peer network. Without background location the app will never get any new information, it will be a pretty boring place. We never keep, store, process, or share your location data with anyone or any device.'
      }
    case 'notifications':
      return {
        title: 'Notifications',
        subtitle: 'Enable notifications',
        details: 'Choose Textile events that trigger notifications. Notifications can be enabled or disabled at any time.'
      }
    case 'receivedInviteNotification':
      return {
        title: 'New Thread Invite',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'deviceAddedNotification':
      return {
        title: 'New Device Paired',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'photoAddedNotification':
      return {
        title: 'New Shared Photos',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'commentAddedNotification':
      return {
        title: 'New Photo Comment',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'likeAddedNotification':
      return {
        title: 'New Photo Like',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'peerJoinedNotification':
      return {
        title: 'Contact Joined Thread',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    case 'peerLeftNotification':
      return {
        title: 'Contact Left Thread',
        subtitle: 'Someone shares a photo with you',
        dependsOn: 'notifications'
      }
    default:
      return undefined
  }
}
