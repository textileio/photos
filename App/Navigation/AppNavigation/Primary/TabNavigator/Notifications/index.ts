import { createStackNavigator } from 'react-navigation'

import Notifications from '../../../../../Containers/FeedList'

import styles, { headerTintColor } from '../../../../Styles/NavigationStyles'

const nav = createStackNavigator(
  {
    Notifications
  },
  {
    headerMode: 'float',
    defaultNavigationOptions: {
      headerStyle: styles.header,
      headerTitleStyle: styles.headerTitle,
      headerTintColor
    }
  }
)

export default nav
