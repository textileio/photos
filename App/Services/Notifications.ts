import { pb, util } from '@textile/react-native-sdk'
import { Notification } from '../Models/Notifications'
import RNPushNotification from 'react-native-push-notification'
import { Alert } from 'react-native'

export interface INotificationsPayload {
  title: string,
  message: string,
  feed: string
  typeString: string,
}

export async function enable(): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      RNPushNotification.requestPermissions()
      resolve()
    } catch (error) {
      reject(error)
    }
  })
}

export function toTypedNotification(notificationData: pb.INotification): Notification {
  const { subjectDesc, subject, target, type, date, user, ...baseNotification } = notificationData
  const d = util.timestampToDate(date)
  switch (type) {
    case pb.Notification.Type.INVITE_RECEIVED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadName: subjectDesc
      }
    case pb.Notification.Type.ACCOUNT_PEER_JOINED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type
      }
    case pb.Notification.Type.PEER_JOINED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc
      }
    case pb.Notification.Type.PEER_LEFT:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc
      }
    case pb.Notification.Type.MESSAGE_ADDED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc,
        target
      }
    case pb.Notification.Type.FILES_ADDED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc,
        target
      }
    case pb.Notification.Type.COMMENT_ADDED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc,
        target
      }
    case pb.Notification.Type.LIKE_ADDED:
      return {
        ...baseNotification,
        date: d,
        username: user.name,
        avatar: user.avatar,
        type,
        threadId: subject,
        threadName: subjectDesc,
        target
      }
  }
}

// Returns blockId if photo type
export function getPhotoId(notification: Notification): string | undefined {
  switch (notification.type) {
    case pb.Notification.Type.COMMENT_ADDED:
    case pb.Notification.Type.FILES_ADDED:
    case pb.Notification.Type.LIKE_ADDED:
      return notification.target
    default:
      return
  }
}

export function notificationTypeToString(type: pb.Notification.Type): string {
  switch (type) {
    case pb.Notification.Type.ACCOUNT_PEER_JOINED:
      return 'ACCOUNT_PEER_JOINED'
    case pb.Notification.Type.COMMENT_ADDED:
      return 'COMMENT_ADDED'
    case pb.Notification.Type.FILES_ADDED:
      return 'FILES_ADDED'
    case pb.Notification.Type.INVITE_RECEIVED:
      return 'INVITE_RECEIVED'
    case pb.Notification.Type.LIKE_ADDED:
      return 'LIKE_ADDED'
    case pb.Notification.Type.MESSAGE_ADDED:
      return 'MESSAGE_ADDED'
    case pb.Notification.Type.PEER_JOINED:
      return 'PEER_JOINED'
    case pb.Notification.Type.PEER_LEFT:
      return 'PEER_LEFT'
  }
}

export function toPayload(notification: Notification): INotificationsPayload {
  const typeString = notificationTypeToString(notification.type)
  const actor = notification.username || 'A contact' // TODO: We want username here, need to look it up?
  switch (notification.type) {
    case(pb.Notification.Type.INVITE_RECEIVED): {
      const title = 'New Invite'
      const message =  [actor, notification.body].join(' ')
      const feed = [actor, notification.body, 'to', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.ACCOUNT_PEER_JOINED): {
      const title = 'New Contact'
      const message = 'You connected to a new account contact'
      const feed = message
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.PEER_JOINED): {
      const title = notification.threadName
      const message =  [actor, notification.body].join(' ')
      const feed = [actor, notification.body, 'group', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.PEER_LEFT): {
      const title = notification.threadName
      const message =  [actor, notification.body].join(' ')
      const feed = [actor, notification.body, 'group', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.MESSAGE_ADDED): {
      const title = notification.threadName
      const message =  [actor, 'wrote', notification.body].join(' ')
      const feed = message
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.FILES_ADDED): {
      const title = notification.threadName
      const body = 'added a photo'
      const message =  [actor, body].join(' ')
      const feed = [actor, body, 'to', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.COMMENT_ADDED): {
      const title =  notification.threadName
      const message = [actor, notification.body].join(' ')
      const body = 'commented on a photo' // todo: fix "a" vs "your"
      const feed = [actor, body, 'in', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
    case(pb.Notification.Type.LIKE_ADDED): {
      const title = notification.threadName
      const body = 'liked a photo' // todo: fix "a" vs "your"
      const message = [actor, body].join(' ')
      const feed = [actor, body, 'in', notification.threadName].join(' ')
      return { title, message, feed, typeString }
    }
  }
}

export async function createNew(notification: Notification): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    try {
      const payload = toPayload(notification)
      if (!payload) {
        return
      }
      RNPushNotification.localNotification({
        title: payload.title,
        message: `${payload.message}.`,
        /* Android Only Property */
        group: payload.typeString, // (optional) add group to message
        /* iOS Only Property */
        category: payload.typeString, // (optional) default: null
        userInfo: { notification },

        /* Android Only Properties */
        largeIcon: 'ic_launcher', // (optional) default: "ic_launcher"
        smallIcon: 'ic_notification', // (optional) default: "ic_notification" with fallback for "ic_launcher"
        // id: '22', // (optional) Valid unique 32 bit integer specified as string. default: Autogenerated Unique ID
        // actions: '["Yes", "No"]',  // (Android only) See the doc for notification actions to know more
        // ticker: "My Notification Ticker", // (optional)
        // bigText: "My big text that will be shown when notification is expanded", // (optional) default: "message" prop
        // subText: "This is a subText", // (optional) default: none

        /* iOS only properties */
        // alertAction: 'view' // (optional) default: view
        playSound: false,
        vibrate: false
      })
      resolve()
    } catch (error) {
      reject()
    }
  })
}

export function displayInvitePromise(message: string): Promise<void> {
  return new Promise<void>((resolve, reject) => {
    Alert.alert(
      'Accept Invite',
      message,
      [
        {
          text: 'Accept',
          onPress: resolve
        },
        {
          text: 'Ignore',
          style: 'cancel',
          onPress: reject
        }
      ],
      { cancelable: false }
    )
  })
}

export async function displayInviteAlert(message: string) {
  await displayInvitePromise(message)
  return true
}
