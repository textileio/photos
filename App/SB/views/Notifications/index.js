import React from 'react'
import { connect } from 'react-redux'
import { View, Text, ScrollView, Platform } from 'react-native'
import PreferencesActions from '../../../Redux/PreferencesRedux'
import PermissionsInfo from '../../components/PermissionsInfo'
import HeaderButtons, { Item } from 'react-navigation-header-buttons'
import SettingsRow from '../../components/SettingsRow'
import GetServiceInfo from './GetServiceInfo'

import { TextileHeaderButtons, Item as TextileItem } from '../../../Components/HeaderButtons'

import styles from './statics/styles'
import Avatar from '../../../Components/Avatar'
import { NavigationActions } from 'react-navigation'

class Notifications extends React.PureComponent {
  constructor (props) {
    super(props)
    this.state = {
      complete: false,
      iOS: Platform.OS === 'ios',
      cameraRoll: false,
      locationBackground: false,
      infoVisible: false,
      info: { }
    }
  }

  static navigationOptions = ({ navigation }) => {
    return {
      headerTitle: 'Notifications',
      headerLeft: (
        <TextileHeaderButtons left>
          <TextileItem title='Back' iconName='arrow-left' onPress={() => { navigation.dispatch(NavigationActions.back()) }} />
        </TextileHeaderButtons>
      ),
      headerRight: (
        <HeaderButtons>
          <Item
            title='Avatar'
            buttonWrapperStyle={{ marginLeft: 11, marginRight: 11 }}
            ButtonElement={
              <Avatar style={{ width: 32, height: 32 }} />
            }
          />
        </HeaderButtons>
      )
    }
  }

  toggleService (name) {
    if (name === 'notifications') {
      // never prompt the user later to get those
      this.props.completeScreen(name)
    }
    this.props.toggleServicesRequest(name)
  }

  hideInfo () {
    this.setState({ infoVisible: false })
  }

  showInfo (service) {
    const info = GetServiceInfo(service)
    this.setState({ infoVisible: true, info })
  }

  render () {
    return (
      <View style={styles.container}>
        <View style={styles.header}>
          <Text style={styles.title}>Choose the types of notifications you want to receive.</Text>
        </View>
        <ScrollView style={styles.contentContainer}>
          <View style={styles.listContainer}>
            {Object.keys(this.props.services)
              .map((service, i) => {
                const value = !!this.props.services[service].status
                let children = Object.keys(this.props.children)
                  .filter((key) => this.props.children[key].info.dependsOn === service)
                  .reduce((previous, current) => {
                    previous[current] = this.props.children[current]
                    return previous
                  }, {})

                return (
                  <View key={i} >
                    <SettingsRow service={service} info={this.props.services[service].info} value={value} infoPress={this.showInfo.bind(this)} onChange={this.toggleService.bind(this)} />
                    {children && Object.keys(children).map((child, i) =>
                      <SettingsRow
                        key={i * 33}
                        child
                        service={child}
                        info={this.props.children[child].info}
                        disabled={!value}
                        value={!!this.props.children[child].status}
                        infoPress={this.showInfo.bind(this)}
                        onChange={this.toggleService.bind(this)}
                      />
                    )}
                  </View>
                )
              }
              )}
          </View>
        </ScrollView>
        {this.state.infoVisible && <PermissionsInfo isVisible info={this.state.info} close={this.hideInfo.bind(this)} />}
      </View>
    )
  }
}

const mapStateToProps = state => {
  // get all top level services
  const allServices = Object.keys(state.preferences.services)
    .reduce((previous, current) => {
      let basic = {
        status: state.preferences.services[current].status
      }
      basic.info = GetServiceInfo(current)
      previous[current] = basic
      return previous
    }, {})

  const services = Object.keys(allServices)
    .filter((key) => allServices[key].info !== undefined && allServices[key].info.dependsOn === undefined)
    .reduce((previous, current) => {
      previous[current] = allServices[current]
      return previous
    }, {})

  // get any services that depend on top level services
  const children = Object.keys(allServices)
    .filter((key) => allServices[key].info !== undefined && allServices[key].info.dependsOn !== undefined)
    .reduce((previous, current) => {
      previous[current] = allServices[current]
      return previous
    }, {})

  return {
    allServices,
    services,
    children
  }
}

const mapDispatchToProps = dispatch => {
  return {
    toggleServicesRequest: (name) => { dispatch(PreferencesActions.toggleServicesRequest(name)) },
    completeScreen: (name) => { dispatch(PreferencesActions.completeTourSuccess(name)) }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Notifications)
