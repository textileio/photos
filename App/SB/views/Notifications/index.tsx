import React from 'react'
import { Dispatch } from 'redux'
import { connect } from 'react-redux'
import { View, Text, ScrollView, Platform } from 'react-native'
import { RootAction, RootState } from '../../../Redux/Types'
import PreferencesActions, { ServiceType, TourScreens } from '../../../Redux/PreferencesRedux'
import PermissionsInfo from '../../components/PermissionsInfo'
import HeaderButtons from 'react-navigation-header-buttons'
import SettingsRow from '../../components/SettingsRow'
import GetServiceInfo from './GetServiceInfo'

import { TextileHeaderButtons, Item as TextileItem } from '../../../Components/HeaderButtons'

import styles from './statics/styles'
import Avatar from '../../../Components/Avatar'
import { NavigationActions, NavigationScreenProps } from 'react-navigation'

interface DispatchProps {
  toggleServicesRequest: (name: ServiceType) => void
  completeScreen: (name: TourScreens) => void
  toggleNodeState: () => void
  toggleNodeErrors: () => void
  toggleNodeStateOverlay: () => void
}

interface ServiceInfo {
  title: string,
  subtitle: string,
  dependsOn?: string
  details?: string
}

interface ServiceSummary {
  [key: string]: {status: boolean, info: ServiceInfo}
}

interface StateProps {
  allServices: ServiceSummary
  services: ServiceSummary
  children: ServiceSummary
  verboseUi: boolean
  verboseUiOptions: {
    nodeStateOverlay: boolean
    nodeStateNotifications: boolean
    nodeErrorNotifications: boolean
  }
}

interface OwnProps {
  complete: boolean
  iOS: boolean
  cameraRoll: boolean
  locationBackground: boolean
  infoVisible: boolean
  info?: ServiceInfo
}

type Props = DispatchProps & StateProps & OwnProps

class Notifications extends React.PureComponent<Props> {
  static navigationOptions = ({ navigation }: NavigationScreenProps<{}>) => {
    return {
      headerTitle: 'Notifications',
      headerLeft: (
        <TextileHeaderButtons left={true}>
          <TextileItem
            title='Back'
            iconName='arrow-left'
            /* tslint:disable-next-line */
            onPress={() => {
              navigation.dispatch(NavigationActions.back())
            }}
          />
        </TextileHeaderButtons>
      ),
      headerRight: (
        <HeaderButtons>
          <HeaderButtons.Item
            title='Avatar'
            buttonWrapperStyle={{ marginLeft: 11, marginRight: 11 }}
            ButtonElement={<Avatar style={{ width: 32, height: 32 }} />}
          />
        </HeaderButtons>
      )
    }
  }

  state: OwnProps = {
    complete: false,
    iOS: Platform.OS === 'ios',
    cameraRoll: false,
    locationBackground: false,
    infoVisible: false
  }

  toggleService = (name: string) => {
    if (name === 'notifications') {
      // never prompt the user later to get those
      this.props.completeScreen(name)
    }
    this.props.toggleServicesRequest(name as ServiceType)
  }

  hideInfo = () => {
    this.setState({ infoVisible: false })
  }

  showInfo = (service: string) => {
    const info = GetServiceInfo(service)
    this.setState({ infoVisible: true, info })
  }

  render () {
    return (
      <View style={styles.container}>
        <View style={styles.header}>
          <Text style={styles.title}>Choose the types of notifications you want to receive.</Text>
        </View>
        <ScrollView style={styles.contentContainer}>
          <View style={styles.listContainer}>
            {Object.keys(this.props.services)
              .map((service, i) => {
                const value = !!this.props.services[service].status
                const children = Object.keys(this.props.children)
                  .filter((key) => this.props.children[key].info.dependsOn === service)
                  .reduce((previous, current) => {
                    previous[current] = this.props.children[current]
                    return previous
                  }, {})

                return (
                  <View key={i} >
                    <SettingsRow
                      service={service}
                      info={this.props.services[service].info}
                      value={value}
                      infoPress={this.showInfo}
                      onChange={this.toggleService}
                    />
                    {children && Object.keys(children).map((child, i) =>
                      <SettingsRow
                        key={i * 33}
                        child={child}
                        service={child}
                        info={this.props.children[child].info}
                        disabled={!value}
                        value={!!this.props.children[child].status}
                        infoPress={this.showInfo}
                        onChange={this.toggleService}
                      />
                    )}
                  </View>
                )
              }
            )}
          </View>

          {this.props.verboseUi &&
            <View style={styles.listContainer}>
              <View>
                <SettingsRow
                  service={'StateOverlay'}
                  info={{title: 'State Overlay'}}
                  value={this.props.verboseUiOptions.nodeStateOverlay}
                  /* tslint:disable-next-line */
                  infoPres={() => {}}
                  onChange={this.props.toggleNodeStateOverlay}
                />
              </View>
              <View>
                <SettingsRow
                  service={'NodeState'}
                  info={{title: 'Node State Change'}}
                  value={this.props.verboseUiOptions.nodeStateNotifications}
                  /* tslint:disable-next-line */
                  infoPres={() => {}}
                  onChange={this.props.toggleNodeState}
                />
              </View>
              <View>
                <SettingsRow
                  service={'NodeErrors'}
                  info={{title: 'Node Errors'}}
                  value={this.props.verboseUiOptions.nodeErrorNotifications}
                  /* tslint:disable-next-line */
                  infoPres={() => {}}
                  onChange={this.props.toggleNodeErrors}
                />
              </View>
            </View>
          }
        </ScrollView>
        {this.state.infoVisible &&
          <PermissionsInfo isVisible={true} info={this.state.info} close={this.hideInfo} />
        }
      </View>
    )
  }
}

const mapStateToProps = (state: RootState): StateProps => {
  // get all top level services
  const allServices = Object.keys(state.preferences.services)
    .reduce((previous, current) => {
      const basic = {
        status: state.preferences.services[current].status,
        info: GetServiceInfo(current)
      }
      previous[current] = basic
      return previous
    }, {})

  const services = Object.keys(allServices)
    .filter((key) => allServices[key].info !== undefined && allServices[key].info.dependsOn === undefined)
    .reduce((previous, current) => {
      previous[current] = allServices[current]
      return previous
    }, {})

  // get any services that depend on top level services
  const children = Object.keys(allServices)
    .filter((key) => allServices[key].info !== undefined && allServices[key].info.dependsOn !== undefined)
    .reduce((previous, current) => {
      previous[current] = allServices[current]
      return previous
    }, {})

  return {
    allServices,
    services,
    children,
    verboseUi: state.preferences.verboseUi,
    verboseUiOptions: state.preferences.verboseUiOptions
  }
}

const mapDispatchToProps = (dispatch: Dispatch<RootAction>): DispatchProps => {
  return {
    toggleServicesRequest: (name) => { dispatch(PreferencesActions.toggleServicesRequest(name)) },
    completeScreen: (name) => { dispatch(PreferencesActions.completeTourSuccess(name)) },
    toggleNodeState: () => { dispatch(PreferencesActions.toggleNodeStateNotifications()) },
    toggleNodeErrors: () => { dispatch(PreferencesActions.toggleNodeErrorNotifications()) },
    toggleNodeStateOverlay: () => { dispatch(PreferencesActions.toggleNodeStateOverlay()) }
  }
}

export default connect(mapStateToProps, mapDispatchToProps)(Notifications)
