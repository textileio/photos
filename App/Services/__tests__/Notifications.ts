import {
    enable,
    toPayload
} from '../Notifications'
import { Notification } from '../../Models/Notifications'
import { pb } from '@textile/react-native-sdk'

describe('notifications', () => {
    describe('enable', () => {
        it('should fire notification lib successfully', async () => {
            await expect(enable()).resolves.toMatchSnapshot()
        })
    })
    describe('toPayload', () => {
        const base = {
            id: 'noteId',
            date: new Date(),
            actor: 'EFG',
            block: 'block',
            read: false,
            username: 'mock test',
            avatar: 'avatar',
            threadId: 'ABC',
            threadName: 'Great Thread',
            target: 'DEF'
        }
        // it('should create InviteReceivedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'invited you to join',
        //         type: pb.Notification.Type.INVITE_RECEIVED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create AccountPeerJoinedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'joined',
        //         type: pb.Notification.Type.ACCOUNT_PEER_JOINED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create PeerJoinedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'joined',
        //         type: pb.Notification.Type.PEER_JOINED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create PeerLeftNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'left',
        //         type: pb.Notification.Type.PEER_LEFT
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create MessageAddedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'here is my message!',
        //         type: pb.Notification.Type.MESSAGE_ADDED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create FilesAddedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'added a photo',
        //         type: pb.Notification.Type.FILES_ADDED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create CommentAddedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'commented on a photo: a camel!',
        //         type: pb.Notification.Type.COMMENT_ADDED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
        // it('should create LikeAddedNotification', () => {
        //     const notification = {
        //         ...base,
        //         body: 'liked a photo',
        //         type: pb.Notification.Type.LIKE_ADDED
        //     }
        //     expect(toPayload(notification as Notification)).toMatchSnapshot()
        // })
    })
})
