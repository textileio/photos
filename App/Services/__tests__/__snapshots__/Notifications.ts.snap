// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`notifications enable should fire notification lib successfully 1`] = `undefined`;
